x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/task/ReindexDataStreamPersistentTaskExecutor.java

```diff
@@ -72,22 +83,112 @@ protected void nodeOperation(AllocatedPersistentTask task, ReindexDataStreamTask
         reindexClient.execute(GetDataStreamAction.INSTANCE, request, ActionListener.wrap(response -> {
             List<GetDataStreamAction.Response.DataStreamInfo> dataStreamInfos = response.getDataStreams();
             if (dataStreamInfos.size() == 1) {
-                List<Index> indices = dataStreamInfos.getFirst().getDataStream().getIndices();
-                List<Index> indicesToBeReindexed = indices.stream()
-                    .filter(getOldIndexVersionPredicate(clusterService.state().metadata()))
-                    .toList();
-                reindexDataStreamTask.setPendingIndicesCount(indicesToBeReindexed.size());
-                for (Index index : indicesToBeReindexed) {
-                    reindexDataStreamTask.incrementInProgressIndicesCount(index.getName());
-                    // TODO This is just a placeholder. This is where the real data stream reindex logic will go
-                    reindexDataStreamTask.reindexSucceeded(index.getName());
+                DataStream dataStream = dataStreamInfos.getFirst().getDataStream();
+                if (getOldIndexVersionPredicate(clusterService.state().metadata()).test(dataStream.getWriteIndex())) {
+                    reindexClient.execute(
+                        RolloverAction.INSTANCE,
+                        new RolloverRequest(sourceDataStream, null),
+                        ActionListener.wrap(
+                            rolloverResponse -> reindexIndices(dataStream, reindexDataStreamTask, reindexClient, sourceDataStream),
+                            e -> completeFailedPersistentTask(reindexDataStreamTask, e)
+                        )
+                    );
+                } else {
+                    reindexIndices(dataStream, reindexDataStreamTask, reindexClient, sourceDataStream);
                 }
-
-                completeSuccessfulPersistentTask(reindexDataStreamTask);
             } else {
                 completeFailedPersistentTask(reindexDataStreamTask, new ElasticsearchException("data stream does not exist"));
             }
-        }, reindexDataStreamTask::markAsFailed));
+        }, exception -> completeFailedPersistentTask(reindexDataStreamTask, exception)));
+    }
+
+    private void reindexIndices(
+        DataStream dataStream,
+        ReindexDataStreamTask reindexDataStreamTask,
+        ExecuteWithHeadersClient reindexClient,
+        String sourceDataStream
+    ) {
+        List<Index> indices = dataStream.getIndices();
+        List<Index> indicesToBeReindexed = indices.stream()
+            .filter(getOldIndexVersionPredicate(clusterService.state().metadata()))
+            .filter(index -> index.getName().equals(dataStream.getWriteIndex().getName()) == false)
```

Comment:
Won't this be redundant after the rollover? I though the write index would have a newer version by this point? Maybe this could be an assert instead?

Reply:
Oh I think you're right, and that's just some placeholder code I had in before I added the rollover step. I'll remove that filter.